### PR TITLE
make include_recipe "::foo" use current cookbook

### PR DESCRIPTION
**Pull request by lamont-granquist**
fixes https://tickets.opscode.com/browse/CHEF-3599

### CODE REVIEW

**Comment by lamont-granquist** [comment]
@opscode/client-core @opscode/client-engineers  and @jtimberman you might like this one.

**Comment by jdmundrawala** [comment]
this is sweet :+1:
I hope people have been versioning cookbooks and locking dependencies correctly

**Comment by jtimberman** [comment]
I'm not super wild about this, I think it may be less clear to users what is being included. That is to say, I don't think I'd use it.

**Comment by lamont-granquist** [comment]
but what about the default recipes which are just platform_family case statements around include_recipe into the same cookbook, and now you can do `include_recipe "::_debian"` and then you can rename the cookbook and you don't have to hack up all the internal references to the cookbook name.

**Comment by jaymzh** [comment]
I *love* this. Scoping for the win.

:+1:

**Comment by jtimberman** [comment]
I don't rename cookbooks though. I get the use of this, but I wouldn't use it myself. A large reason for that is backwards compatibility. Though I keep my systems updated with the latest chef obviously not everyone does. 

To be clear, I'm not against this change, but I'm not for it either. :smile_cat:.

**Comment by jonlives** [comment]
I'm also a little concerned about adding confusion here - in theory, all three of the following would now be valid:

include_recipe "foo::default"
include_recipe "foo"
include_recipe "::default"

I quite like the underlying idea, but I'm worried this is going to make things a lot less friendly for less experienced users.

**Comment by jtimberman** [comment]
A bit more on the backwards compatibility, it means any cookbook that uses this syntax won't work on older versions of chef. That sucks. We have seen a lot of time and effort in adding guards around `use_inline_resources` in cookbook providers, because even after two years of chef 11 being available, people are still using 10. 

This feels like a major version change feature to me. It's opt in but the user experience seems like it could be really bad.

**Comment by jtimberman** [comment]
It also feels like this is something we should have an RFC about.

File: lib/chef/recipe.rb [hunk 1]
```diff
@@ -54,12 +54,16 @@ class Recipe
...
-    def self.parse_recipe_name(recipe_name)
-      rmatch = recipe_name.match(/(.+?)::(.+)/)
-      if rmatch
-        [ rmatch[1].to_sym, rmatch[2] ]
+    def self.parse_recipe_name(recipe_name, current_cookbook: nil)
+      case recipe_name
+      when /(.+?)::(.+)/
+        [ $1.to_sym, $2 ]
+      when /^::(.+)/
+        raise "current_cookbook is nil, cannot resolve #{recipe_name}" if current_cookbook.nil?
```
Comment (tyler-ball) [inline lib/chef/recipe.rb]: There should be a unit test for this error being raised

**Comment by jonlives** [comment]
Seconded, given this breaks backwards compat (which i didn't think of), I'd say an RFC might be a good idea.

**Comment by lamont-granquist** [comment]
We can't make changes like this be major version changes, that's really abusive use of semver.  This is clearly extending the API and old cookbooks work fine and cookbooks that use the new feature will break, but that's the definition of a minor version bump.  The solution for cookbooks that need to support chef 11 or whatever is simply not to use this feature (and chef-10 support and the use_inline_resources hack should have been dropped a long time ago in cookbooks, but we supported chef-10 for nearly two years).

I'm also not sure why this elevates to the level of an RFC, that also seems somewhat abusive of that process.  This is fairly trivial.

**Comment by lamont-granquist** [comment]
We also got a customer request for this previously and triaged it and were going to do it:

https://tickets.opscode.com/browse/CHEF-3599

**Comment by stevendanna** [comment]
@lamont-granquist I'm not sure on the name for it, but if a cookbook works against 12.x but doesn't work against 12.x-1, I agree that it isn't a breaking change, but it does kinda suck until we have a way to declare the version of chef a cookbook works with. 

Anyway, if we are going to have this feature, I would like to see a test covering the following case:

- A recipe "recipeA::default" calls `include_recipe "recipeB::foo" `. recipeB::foo calls `include_recipe "::baz"`

My assumption is that this should result in "recipeB::baz" being called.  Looking at the code it appears that this works currently, but I'd like a test covering that behavior since it feels like the sort of thing that could break in the future.

**Comment by tyler-ball** [comment]
From a code quality perspective, I'm :+1: with a minor comment.

It seems like the only way this breaks backwards compatibility is with public cookbooks.  Does the Supermarket have a way of marking a cookbook as only working with specific versions of Chef?

**Comment by lamont-granquist** [comment]
Can't do that now, its been often requested:

https://tickets.opscode.com/browse/CHEF-3591
https://tickets.opscode.com/browse/CHEF-2251
https://tickets.opscode.com/browse/CHEF-4524

**Comment by ranjib** [comment]
@jtimberman im not sure if this breaks backward compatibility. Older recipes will run just fine, new recipes that will use this feature wont run on old chef versions, and thats fine. This is new. Backward compatibility means older stuff should run intact. This preserve the old interface.

**Comment by stevendanna** [comment]
@ranjib @lamont-granquist @jtimberman @jonlives Perhaps there is a discussion to be had outside of this particular issue on the projects view of semver, breaking changes, etc.  For instance, if we were to nak this, we'd have to nak most DSL improvements that don't happen in a major version.  That seems ill-advised.

**Comment by danielsdeleo** [comment]
Agree. For me, compatibility means you can upgrade chef-client and your existing cookbooks work. Individual cookbook projects will have to decide for themselves if changing the minimum required chef version constitutes a backwards incompatible change and when to start relying on new features, but that's outside the scope of what a compat breaking change is for chef-client itself.

**Comment by jtimberman** [comment]
Maybe I'm overly cautious and being too prescriptive about my opinion that this is a compatibility issue. That said, am I missing something in thinking this may end up in negative user experience with the product?

To me it does feel like a fairly major change, which is why I suggested that we have an RFC for it. Am I missing something there?

**Comment by lamont-granquist** [comment]
it feels completely trivial to me, and kind of a bit of obvious syntactic sugar, and this seems like a bikeshed and i don't see how making it into an RFC will shine any more light onto it.

**Comment by jonlives** [comment]
I'm not sure I entirely agree this is a bikeshed...reading through the above comments, my thoughts are currently that it's not ideal to potentially have cookbooks (and it sound like people will already want to use this) that will work with eg. chef-client 12.0.4 but not 12.0.3 - my worry is that will be rather confusing for anybody who ends up using that cookbook.

Do we currently have a way in Chef Supermarket to indicate the minimum required chef-client version? My objections here are really more user experience - I like the underlying idea, but this feels odd to break back-compat in a minor release unless we have a way to ensure people can avoid it.

**Comment by lamont-granquist** [comment]
So we currently don't have a mechanism for cookbooks to indicate that they're using a feature that was introduced in a minor release, and we release features in minor releases all the time.  That is the definition of a minor release.   Cookbooks that pick up the new features will be broken on older versions of chef client.  If we push all those features into major releases, then you literally make minor releases pointless and reduce chef-client development to just constant patching and then scrambling to add features every major release.

And some current PRs that would also be affected:

#2770:  cookbooks that use http/ftp sources on rpm_package would be broken in 12.0.0
#2767:  cookbooks that use allow_downgrade in rpm_package would be broken in 12.0.0
#2692:  passing arrays to yum+deb providers would be broken in 12.0.0
#2581:  using file validation features would be broken in 12.0.0

If you go down this road you can't add any flags to resources except outside of a new major release.

I agree we should add metadata to be able to have cookbooks indicate which chef-client versions they support, but we've always added new features in minor releases, and then had the problem that opscode-cookbooks has had to lag behind (and forcing features to major releases don't solve this problem, it just makes chef development worse).

We can't do this.

**Comment by jonlives** [comment]
This is a good point @lamont-granquist...I'm not saying I agree with it, but it seems like this is already an established pattern.

With *that* said, the examples you gave relate to specific resources or resource parameters, and we're talking here about entire incompatible cookbooks. 

Possibly a semantic distinction, and I'm prepared to defer to general consensus on this one but it feels to me like this is a feature that, once documented, is likely to be pretty widely used, with the resultant flood of people trying to use community cookbooks with 12.0.less-than-whatever that they had no way of telling would break.

**Comment by jonlives** [comment]
What say ye on this, other @opscode/client-core that haven't already spoken?

**Comment by lamont-granquist** [comment]
"we're talking here about entire incompatible cookbooks" makes no sense at all to me.  If you include an invalid parameter in a resource in a cookbook then the cookbook will fail validation on old chef-client versions and will break the whole cookbook.

If you look at this you just gave a :+1: to an RFC that would add additional sugar to the DSL as well and would cause cookbooks that used the new feature to break on chef-client 12.0.0:  opscode/chef-rfc#83

I'm really hugely frustrated with the amount of RFC and PRs that this would block and why this one is getting dogpiled on over this issue.  Just got and look *clearly* at the all the other work that this idea would block.

**Comment by jonlives** [comment]
@lamont-granquist we're not in a place where a single dissenting maintainer can block this PR - majority is in favour, and there's no veto from higher up. 

My sole concern here is that people are going to be downloading cookbooks from supermarket that have hidden version requirements, and this just happens to be the PR I'm expressing that thought in.  Please realise, my feedback is not intended to dogpile this issue, or you, and I don't plan to stand in the way of its implementation.

I guess I just don't want us to be too cavalier about giving users a less-optimal user experience because that's how we already do it, or because it slows momentum not to.

I think the reason that hit me here is because we're dealing with a core DSL method rather than a specific resource - as you say, this isn't all that different in functional terms.

**Comment by lamont-granquist** [comment]
Okay, then this discussion needs to be taken off to some other venue.  Since the PR comments were addressed, I'll be merging this.

**Comment by jonlives** [comment]
Yeah - I'll write up an RFC on adding something like we talked about above to Supermarket.

**Comment by jonlives** [comment]
:+1:

**Pull request merged**